File: Trees/Python/Binary Tree Level Order Traversal II.py
# Problem link: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/description/
# Time: O(n)
# Space: O(n)# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        result, queue = [], deque()

        if root:
            queue.append(root)
        
        while queue:
            temp = []

            for _ in range(len(queue)):
                node = queue.popleft()
                temp.append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            result.append(temp)

        return result[::-1]

File: Trees/Java/Binary Tree Level Order Traversal II.java
// Problem link: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/description/
// Time: O(n)
// Space: O(n)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();

        if(root != null) {
            queue.offer(root);
        }

        while(!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> temp = new ArrayList<>();

            for(int idx = 0; idx < size; idx++) {
                TreeNode node = queue.poll();
                temp.add(node.val);

                if(node.left != null) {
                    queue.offer(node.left);
                } 
                if(node.right != null) {
                    queue.offer(node.right);
                }
            }

            result.add(temp);
        }

        Collections.reverse(result);
        return result;
    }
}
